src/uct/ib/mlx5/gdaki/gdaki.c

```diff
@@ -722,11 +726,140 @@ uct_gdaki_md_check_uar(uct_ib_mlx5_md_t *md, CUdevice cuda_dev)
     return status;
 }
 
+typedef struct {
+    unsigned               index;
+    ucs_sys_dev_distance_t dist;
+    int                    usecount;
+} uct_gdaki_dev_score_t;
+
+typedef struct {
+    ucs_sys_device_t sys_dev;
+    size_t           num;
+    CUdevice         gpus[UCT_GDAKI_MAX_GPU_PER_IB];
+} uct_gdaki_dev_matrix_elem_t;
+
+static uct_gdaki_dev_matrix_elem_t *uct_gdaki_dev_matrix;
+static size_t uct_gdaki_dev_matrix_length;
+
+static int uct_gdaki_dev_matrix_score(const void *pa, const void *pb, void *arg)
+{
+    const uct_gdaki_dev_score_t *a = pa;
+    const uct_gdaki_dev_score_t *b = pb;
+    int res = ucs_score_cmp(a->dist.latency, b->dist.latency);
+    return res ? res : ucs_signum(a->usecount - b->usecount);
+}
+
+static ucs_status_t uct_gdaki_dev_matrix_init(unsigned max_ib_per_gpu)
+{
+    ucs_status_t status = UCS_OK;
+    int i, j, num_ibs, num_gpus;
+    struct ibv_device **device_list;
+    struct ibv_device *ib_dev;
+    uct_gdaki_dev_score_t *scores;
+    char *path_buffer;
+    ucs_sys_device_t gpu_sys_dev;
+    const char *sysfs_path;
+    unsigned ib_per_gpu;
+    uct_gdaki_dev_matrix_elem_t *ibdesc;
+    CUdevice gpu_dev;
+
+    status = ucs_string_alloc_path_buffer(&path_buffer, "path_buffer");
+    if (status != UCS_OK) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    device_list = ibv_get_device_list(&num_ibs);
+    if (device_list == NULL) {
+        status = UCS_ERR_IO_ERROR;
+        goto out_buff;
+    }
+
+    ucs_assert(num_ibs > 0);
+    uct_gdaki_dev_matrix = ucs_calloc(num_ibs, sizeof(*uct_gdaki_dev_matrix),
```

Comment:
IMO it would be better that this function allocates and returns the allocated matrix, instead of directly modifying global variables, because it does not directly take a global lock